File: GEE_precipitation.js
// -------------------------------------------
// 1. Load your lat/lon CSV into GEE
// -------------------------------------------
// Upload your CSV to GEE Assets as a Table
// Replace asset ID below with yours
var islands = ee.FeatureCollection('users/yourusername/ime_crep_lat_lon')
  .map(function(f) {
    return f.setGeometry(ee.Geometry.Point([f.get('lon'), f.get('lat')]));
  });

// -------------------------------------------
// 2. Load CHIRPS daily and convert to monthly totals
// -------------------------------------------
var chirps = ee.ImageCollection('UCSB-CHG/CHIRPS/DAILY')
  .filterDate('1991-01-01', '2020-12-31');

// Function to sum a month for each year, then take the average over years
function monthMean(m) {
  m = ee.Number(m);
  var monthly = chirps.filter(ee.Filter.calendarRange(m, m, 'month'))
    .map(function(img) {
      return img.set('year', img.date().get('year'));
    })
    .aggregate_array('year'); // for debugging

  var perYear = chirps.filter(ee.Filter.calendarRange(m, m, 'month'))
    .reduce(ee.Reducer.sum()); // sum over all years
  
  var climatology = chirps.filter(ee.Filter.calendarRange(m, m, 'month'))
    .map(function(img) {
      return img.set('year', img.date().get('year'));
    })
    .reduce(ee.Reducer.mean()); // mean across all days
  
  // Simpler method: sum daily, then divide by number of years
  var monthSumAllYears = chirps.filter(ee.Filter.calendarRange(m, m, 'month'))
    .sum();
  var meanMonth = monthSumAllYears.divide(30); // 30 years in 1991–2020
  
  return meanMonth.set('month', m);
}

// List of months and abbreviations
var monthNums = ee.List.sequence(1, 12);
var monthAbbr = ee.List(['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec']);

// Get monthly climatology images
var monthlyImages = monthNums.map(monthMean);

// -------------------------------------------
// 3. Sample at island points and combine into wide format
// -------------------------------------------
var results = islands.map(function(f) {
  var props = ee.Dictionary();
  
  // Add each month as a property
  monthNums.zip(monthAbbr).iterate(function(mPair, acc) {
    mPair = ee.List(mPair);
    var mNum = mPair.get(0);
    var mName = ee.String(mPair.get(1));
    
    var monthImg = ee.Image(monthlyImages.get(ee.Number(mNum).subtract(1)));
    var val = monthImg.reduceRegion({
      geometry: f.geometry(),
      reducer: ee.Reducer.mean(),
      scale: 5000
    }).values().get(0);
    
    acc = ee.Dictionary(acc).set(mName, val);
    return acc;
  }, props);
  
  // Compute annual mean (mean of monthly means)
  var monthlyVals = monthAbbr.map(function(m) {
    return ee.Number(ee.Dictionary(props).get(ee.String(m)));
  });
  var annualMean = ee.Array(monthlyVals).reduce('mean', [0]).get([0]);
  
  return f.set(props).set('AnnualMean', annualMean);
});

// -------------------------------------------
// 4. Export as CSV
// -------------------------------------------
Export.table.toDrive({
  collection: results,
  description: 'CHIRPS_1991_2020_MonthlyClimatology_Wide',
  fileFormat: 'CSV'
});

File: gee_precipitation.R
library(rgee)
# Initialize Earth Engine
ee_Initialize(drive = TRUE,gcs = FALSE)


library(tidyverse)
library(sf)
library(purrr)


# Convert to EE FeatureCollection
islands_ee <- sf_as_ee(isl_dat %>% mutate(lat = latitude, lon = longitude) %>% select(geometry, island, lat, lon))

# -------------------------
# 3. Load CHIRPS daily data
# -------------------------
chirps <- ee$ImageCollection("UCSB-CHG/CHIRPS/DAILY") %>%
  ee$ImageCollection$filterDate('2000-01-01', '2020-12-31') %>% 
  ee$ImageCollection$filterBounds(islands_ee)

# -------------------------
# 4. Compute average total monthly precipitation per month
# -------------------------
start_year <- 2000
end_year   <- 2020

# ---- Safe monthly-total (sum of daily) image maker
monthly_total_img <- ee_utils_pyfunc(function(ym) {
  d <- ee$Dictionary(ym)
  y <- ee$Number(d$get('y'))
  m <- ee$Number(d$get('m'))
  
  start <- ee$Date$fromYMD(y, m, 1)
  end   <- start$advance(1, 'month')
  
  # Filter CHIRPS for this month and sum daily precipitation
  img <- chirps$filterDate(start, end)$select('precipitation')$sum()
  
  # Set month/year properties
  img$set('year', y)$set('month', m)
})


# ---- Build server-side list of (year,month)
ym_list <- ee$List$sequence(start_year, end_year)$map(
  ee_utils_pyfunc(function(y)
    ee$List$sequence(1,12)$map(
      ee_utils_pyfunc(function(m) ee$Dictionary(list(y=y, m=m)))
    )
  )
)$flatten()

monthly_ic <- ee$ImageCollection$fromImages(ym_list$map(monthly_total_img))

# ---- Zonal stats per island: mean across pixels (mm/month per island)
# NOTE: if you truly want area-scaled totals, change Reducer$mean() → Reducer$sum()
island_monthly_fc <- monthly_ic$map(
  ee_utils_pyfunc(function(img) {
    
    # Reduce over islands to get mean and pixel count
    reduced <- img$reduceRegions(
      collection = islands_ee,
      reducer = ee$Reducer$mean()$combine(
        reducer2 = ee$Reducer$count(),
        sharedInputs = TRUE
      ),
      scale = 5566,   # CHIRPS resolution ~5.56 km
      tileScale = 2
    )
    
    # Add year and month properties to each island
    reduced$map(
      ee_utils_pyfunc(function(f) {
        f$set('year', img$get('year'))$
          set('month', img$get('month'))
      })
    )
    
  })
)$flatten()

# Export to Google Drive
task <- ee_table_to_drive(
  collection = island_monthly_fc,
  description = 'island_monthly_precip',
  folder = 'GEE_exports',
  fileFormat = 'CSV'
)

task$start()
ee_monitoring()

# data copied from Drive into repo 
island_monthly_df <- read.csv('data/gee-exports/island_monthly_precip_2025_08_22_12_30_23.csv') %>% 
  select(count, island, lat, lon, mean, month, year) %>% 
  mutate(
    year  = as.integer(year),
    month = as.integer(month),
    monthly_total_mm = mean,   # mean of monthly totals across pixels
    n_pixel = count   # number of pixels
  ) %>%
  select(-count, -mean) %>% 
  arrange(island, year, month)


# local side version
# island_monthly_df <- ee_as_sf(island_monthly_fc) %>% 
#   transmute(
#     year  = as.integer(year),
#     month = as.integer(month),
#     monthly_total_mm = mean,   # mean of monthly totals across pixels
#     n_pixel = count   # number of pixels
#   ) %>%
#   arrange(island, year, month)

# ---- Monthly climatology (avg across years). n = 26 islands.
clim_monthly <- island_monthly_df %>%
  filter(!n_pixel == 0) %>% # This drops Baker, Laysan, Maug, Necker, Nihoa
  group_by(island, month, n_pixel) %>%
  summarize(avg_monthly_mm = mean(monthly_total_mm, na.rm = TRUE)) %>%
  mutate(month_label = factor(month, levels = 1:12, labels = month.abb)) %>%
  select(island, n_pixel, month, month_label, avg_monthly_mm)

# Baker has NA precipitation because they do not overlap with a pixel (ie islands too small / land area not given pixel)
write.csv(clim_monthly, file = 'data/gee-exports/crep_monthly_precipitation_mm.csv', row.names=FALSE)
